File: frontend/src/types/request.ts
// import { type User } from  "./user";

export type TFullRequest = {
  id: number;
  status: TStatus;
  moving_date: string | null;
  service_id: number;
  packing_id: number;
  service?: {
    id: number;
    name: string;
  };
  packing?: {
    id: number;
    name: string;
  };
  work_time: {
    min: number;
    max: number;
  };
  total_time: {
    min: number;
    max: number;
  };
  total_price: {
    min: number;
    max: number;
  };
  min_total_time: number;
  details: {
    delicate_items_question_answer?: string;
    bulky_items_question_answer?: string;
    disassemble_items_question_answer?: string;
    comments?: string;
    is_touched?: boolean;
  };
  size: string;
  start_time_window: number | null;
  end_time_window: number | null;
  travel_time: number;
  crew_size: number;
  rate: number;
  paired_request_id: number | null;
  paired_request: TFullRequest | null;
  is_moving_from_storage: boolean;
  customer: TCustomer | null;
  customer_id: number;
  foreman_id: number | null,
  mover_ids: number[]
  origin: TAddress;
  destination: TAddress;
  stops: TStop[];
  sales_notes: string;
  driver_notes: string;
  customer_notes: string;
  dispatch_notes: string;
  deposit: number;
  can_edit_request: boolean;
  truck_ids: number[];
  image_urls: {
    id: number;
    url: string;
  }[];
  unread_messages_count: number;
  created_at: string;
  updated_at: string;
};
export type TStatus = "pending" | "pending_info" | "pending_date" | "hold" | "not_confirmed" | "confirmed" | "not_available" | "completed" | "spam" | "canceled" | "refused" | "closed" | "expired" | "archived";

export type TAddress = {
  street: string;
  city: string;
  state: string;
  zip: string;
  floor: string;
  apt?: string | undefined;
  location: {
    lat: number;
    lng: number;
  };
}

export type TStop = TAddress & {
  type: string;
};

export type TCustomer = {
  id: number;
  first_name: string;
  last_name: string;
  email: string;
  phone: string;
  additional_phone?: string | undefined;
  additional_email?: string | undefined;
};

export type TMenuRequest = {
  id: number;
  status: TStatus;
  origin: TAddress;
  destination: TAddress;
};

File: app/models/request.rb
class Request < ApplicationRecord
  enum :status,
       %i[
         pending
         pending_info
         pending_date
         hold
         not_confirmed
         confirmed
         not_available
         completed
         spam
         canceled
         refused
         closed
         expired
         archived
       ]

  belongs_to :customer,
             class_name: "User",
             foreign_key: "customer_id",
             optional: true
  belongs_to :foreman,
             class_name: "User",
             foreign_key: "foreman_id",
             optional: true
  has_many :mover_requests
  has_many :movers, through: :mover_requests, source: :user
  has_many_attached :images
  has_and_belongs_to_many :trucks
  belongs_to :service
  belongs_to :packing
  belongs_to :move_size, optional: true
  belongs_to :paired_request, class_name: "Request", optional: true
  # has_one :contract, dependent: :destroy
  # has_many :payments, as: :entity, class_name: "Payment"
  # has_many :messages, dependent: :destroy

  # validates :service, presence: true
  # validates :packing, presence: true
  # validates :moving_date, presence: true
  # validates :status, presence: true
  # validates :size, presence: true
  # validates :crew_size, presence: true
  # validates :rate, presence: true

  # Entrance type methods
  # def origin_entrance_type
  #   return nil unless origin["floor"]&.dig("entrance_type_id")
  #   EntranceType.find_by(id: origin["floor"]["entrance_type_id"])
  # end

  # def destination_entrance_type
  #   return nil unless destination["floor"]&.dig("entrance_type_id")
  #   EntranceType.find_by(id: destination["floor"]["entrance_type_id"])
  # end

  # def origin_entrance_type=(entrance_type)
  #   origin["floor"] ||= {}
  #   origin["floor"]["entrance_type_id"] = entrance_type&.id
  #   save
  # end

  # def destination_entrance_type=(entrance_type)
  #   destination["floor"] ||= {}
  #   destination["floor"]["entrance_type_id"] = entrance_type&.id
  #   save
  # end

  # def origin_floor_value
  #   origin["floor"]&.dig("value")
  # end

  # def destination_floor_value
  #   destination["floor"]&.dig("value")
  # end

  # def origin_floor_value=(value)
  #   origin["floor"] ||= {}
  #   origin["floor"]["value"] = value
  #   save
  # end

  # def destination_floor_value=(value)
  #   destination["floor"] ||= {}
  #   destination["floor"]["value"] = value
  #   save
  # end
end
